contracts/AMM/LiquidityPool/ThalesAMMLiquidityPool.sol

```diff
@@ -0,0 +1,716 @@
+        address roundPool = roundPools[round];
+        // final balance is the final amount of sUSD in the round pool
+        uint currentBalance = sUSD.balanceOf(roundPool);
+        // calculate PnL
+
+        // if no allocation for current round
+        if (allocationPerRound[round] == 0) {
+            profitAndLossPerRound[round] = 1;
+        } else {
+            profitAndLossPerRound[round] = (currentBalance * ONE) / allocationPerRound[round];
+        }
+
+        for (uint i = 0; i < usersPerRound[round].length; i++) {
+            address user = usersPerRound[round][i];
+            uint balanceAfterCurRound = (balancesPerRound[round][user] * profitAndLossPerRound[round]) / ONE;
+            if (userInRound[round][user]) {
+                if (!withdrawalRequested[user] && (profitAndLossPerRound[round] > 0)) {
+                    balancesPerRound[round + 1][user] = balancesPerRound[round + 1][user] + balanceAfterCurRound;
+                    userInRound[round + 1][user] = true;
+                    usersPerRound[round + 1].push(user);
+                    if (address(stakingThales) != address(0)) {
+                        stakingThales.updateVolume(user, balanceAfterCurRound);
+                    }
+                } else {
+                    balancesPerRound[round + 1][user] = 0;
+                    sUSD.safeTransferFrom(roundPool, user, balanceAfterCurRound);
+                    withdrawalRequested[user] = false;
+                    userInRound[round + 1][user] = false;
+                    emit Claimed(user, balanceAfterCurRound);
+                }
+            }
+        }
+
+        //always claim for defaultLiquidityProvider
+        if (balancesPerRound[round][defaultLiquidityProvider] > 0) {
+            uint balanceAfterCurRound = (balancesPerRound[round][defaultLiquidityProvider] * profitAndLossPerRound[round]) /
+                ONE;
+            sUSD.safeTransferFrom(roundPool, defaultLiquidityProvider, balanceAfterCurRound);
+            emit Claimed(defaultLiquidityProvider, balanceAfterCurRound);
+        }
+
+        if (round == 1) {
+            cumulativeProfitAndLoss[round] = profitAndLossPerRound[round];
+        } else {
+            cumulativeProfitAndLoss[round] = (cumulativeProfitAndLoss[round - 1] * profitAndLossPerRound[round]) / ONE;
+        }
+
+        // start next round
+        round += 1;
+
+        //add all carried over sUSD
+        allocationPerRound[round] += sUSD.balanceOf(roundPool);
+
+        totalDeposited = allocationPerRound[round] - balancesPerRound[round][defaultLiquidityProvider];
+
+        address roundPoolNewRound = _getOrCreateRoundPool(round);
+
+        sUSD.safeTransferFrom(roundPool, roundPoolNewRound, sUSD.balanceOf(roundPool));
+
+        emit RoundClosed(round - 1, profitAndLossPerRound[round - 1]);
+    }
+
+    /// @notice Iterate all markets in the current round and exercise those ready to be exercised
+    function exerciseMarketsReadyToExercised() public {
```

Comment:
We should add batching here, because for AMM markets they all end at the same time, so we will have like 100 markets to exercise in parallel, which is likely to breach gas limits